### DISCUSSION TITLE
Initialise object containing an attribute with abstract class typehint from a dict

### DISCUSSION THREAD

**Discussion by jsagild** [pydantic/pydantic#7995] [Question]
Hi! 👋 

I'm trying to initialize an object from a dict using `model_validate` where one attribute is type hinted with an abstract class. I've seen similar discussions, like: #3091 and #5785. AFAIU, the solutions discussed use `animal: Union[Dog, Cat]` where I'd like to use `animal: Animal` where `Animal` is the abstract class. 

Here is a working example of what I want to achieve:
```
import abc
from pydantic import BaseModel, ValidationError, field_validator


class Animal(abc.ABC, BaseModel):
    animal_type: str
    name: str

    @abc.abstractmethod
    def get_sound(self):
        raise NotImplementedError


class Dog(Animal):
    animal_type: str = "Dog"
    breed: str

    def get_sound(self):
        return "woof"


class Cat(Animal):
    animal_type: str = "Cat"
    favorite_toy: str

    def get_sound(self):
        return "meow"


class Owner(BaseModel):
    animal: Animal

    @field_validator("animal", mode="before")
    def validate(cls, v):
        if isinstance(v, dict):
            animal_type = v.get("animal_type")
            name = v.get("name")
            if name is None:
                raise ValidationError("name is required")
            if animal_type == "Dog":
                breed = v.get("breed")
                return Dog(animal_type=animal_type, name=name, breed=breed)
            elif animal_type == "Cat":
                favorite_toy = v.get("favorite_toy")
                return Cat(
                    animal_type=animal_type, name=name, favorite_toy=favorite_toy
                )
            else:
                raise ValidationError(f"Unsupported animal type: {animal_type}")
        return v


owner_1_data = {
    "animal": {
        "animal_type": "Cat",
        "name": "Caesar",
        "favorite_toy": "ball",
    }
}

owner_2_data = {
    "animal": {
        "animal_type": "Dog",
        "name": "Rex",
        "breed": "German Shepherd",
    }
}

owner_1 = Owner.model_validate(owner_1_data)
owner_2 = Owner.model_validate(owner_2_data)

print(owner_1)
print(owner_2)
```
with output
```
animal=Cat(animal_type='Cat', name='Caesar', favorite_toy='ball')
animal=Dog(animal_type='Dog', name='Rex', breed='German Shepherd')
```

### Question
This doesn't feel very right.
- Is there a way to handle this without overwriting the initialization? (And still use `animal: Animal`)
- Is there a "Pydantic" way of doing this?

I would appreciate any help or input! 

### Versions
pydantic==2.1.1
python==3.9.1

**Comment by sydney-runkle** [comment]
Hi @jsagild,

I think the most `pydantic` like solution for this problem would be to use a discriminated `Union`. Here's the docs on those: https://docs.pydantic.dev/dev/concepts/unions/#discriminated-unions.

This suggestion might not meet your criteria though, as this would require the use of `Union[Cat, Dog]`.

**Reply by jsagild** [reply]
Hi @sydney-runkle,
Thank you very much for the answer. 

I see. In my use case I would prefer to use the abstract class when type hinting, and therefore not use the `Union[Cat, Dog]` solution. I'm waiting to mark this as the answer as I'd like to see whether there are other solutions. Thanks again!
